walker.go

```diff
@@ -0,0 +1,429 @@
+	// of the `ActiveNode` is inserted after it (effectively becoming the new
+	// `ActiveNode`).
+	//
+	// The slice must *always* have a length bigger than zero with the root
+	// of the DAG at the first position (empty DAGs are not valid).
+	path []NavigableNode
+
+	// Depth of the `ActiveNode`. It grows downwards, root being 0, its child 1,
+	// and so on. It controls the effective length of `path` and `childIndex`.
+	//
+	// A currentDepth of -1 signals the start case of a new `Walker` that hasn't
+	// moved yet. Although this state is an invalid index to the slices, it
+	// allows to centralize all the visit calls in the `down` move (starting at
+	// zero would require a special visit case inside every walk operation like
+	// `Iterate()` and `Search`). This value should never be returned to after
+	// the first `down` movement, moving up from the root should always return
+	// `errUpOnRoot`.
+	currentDepth int
+
+	// This slice has the index of the child each node in `path` is pointing
+	// to. The child index in the node can be set past all of its child nodes
+	// (having a value equal to `ChildTotal`) to signal it has visited (or
+	// skipped) all of them. A leaf node with no children that has its index
+	// in zero would also comply with this format.
+	//
+	// Complement to `path`, not only do we need to know which nodes have been
+	// traversed to reach the `ActiveNode` but also which child nodes they are
+	// to correctly have the active path of the DAG. (Reword this paragraph.)
+	childIndex []uint
+
+	// Flag to signal that a pause in the current walk operation has been
+	// requested by the user inside `Visitor`.
+	pauseRequested bool
+
+	// Used to pass information from the central `Walker` structure to the
+	// distributed `NavigableNode`s (to have a centralized configuration
+	// structure to control the behavior of all of them), e.g., to tell
+	// the `NavigableIPLDNode` which context should be used to load node
+	// promises (but this could later be used in more elaborate ways).
+	ctx context.Context
+}
+
+// Function called each time a node is arrived upon in a walk operation
+// through the `down` method (not when going back `up`). It is the main
+// API to implement DAG functionality (e.g., read and seek a file DAG)
+// on top of the `Walker` structure.
+//
+// Its argument is the current `node` being visited (the `ActiveNode`).
+// Any error it returns (apart from the internal `errPauseWalkOperation`)
+// will be forwarded to the caller of the walk operation (pausing it).
+//
+// Any of the exported methods of this API should be allowed to be called
+// from within this method, e.g., `Next`.
+// TODO: Check that. Can `ResetPosition` be called without breaking
+// the `Walker` integrity?
+type Visitor func(node NavigableNode) error
+
+// NavigableNode is the interface the nodes of a DAG need to implement in
+// order to be traversed by the `Walker`.
+type NavigableNode interface {
```

Comment:
I see that this lets you move a lot of preloading logic out of the walker, which is nice, but is it possible you are optimizing for a more abstract use case that we might not need, or might not need for a long time? Adding this interface makes usage of the `Walker` more complicated for any normal IPLD code:

```go
// Box my node in a NavigableNode
boxedRoot = NewNavigableIPLDNode(root, dagService)
walker = NewWalker(ctx, boxedRoot)
walker.iterate(func(boxedNode NavigableNode) error {
    // Unbox my node so I can use it
    node = ExtractIPLDNode(boxedNode)
    // Do something
})

// vs.
walker = NewWalker(ctx, root)
walker.iterate(func(node Node) error {
    // Do something
})
```

Reply:
> but is it possible you are optimizing for a more abstract use case that we might not need, or might not need for a long time?

The idea of the walker is to use it (or evaluate using it) not only in the IPLD node case, this can operate on Gx dependencies, MFS directories, HAMT shards, etc.

> Adding this interface makes usage of the `Walker` more complicated for any normal IPLD code:

Yes, since `Node` is already an interface I don't see a cleaner way to do this (maybe @Stebalien can help here). For other entities represented with concrete structures, like a Gx `Dependecy`, we need to only add the `NavigableNode` methods to it and we will transparently be able to use it in the `Visitor` functions without any weird conversion (just type-casting it to `Dependecy`).

Reply:
> The idea of the walker is to use it (or evaluate using it) not only in the IPLD node case, this can operate on Gx dependencies, MFS directories, HAMT shards, etc.

Right, that was clear in your description of the PR. I guess what I was trying to ask was: are those actual, real use cases you *need* to solve here, or just theoretical ones you envision because they feel similar?

The answer might be yes, they’re real! But I’ve also seen plenty of code that is over-complicated because it was trying to solve a set of theoretical use cases that nobody actually needed it to and that it was never put into practice for.

Reply:
Good point. We definitely need a more general description of what a node is (not an IPLD node), hence this `NavigableNode` interface, not only for the walker but for other async fetch operations (now being implemented in the HAMT). But no, this hasn't been tested anywhere else (I need to get this passed first), so it's something to keep in mind. Over-complicating this would make it harder to use elsewhere, when I created the `NavigableNode` I started adding too many methods that I ended up removing (or commenting) to take it now to this reduced 2 method version: tell me how many children you have and fetch the children X.